File: turtle-soup-bot/src/main/kotlin/io/github/kapu/turtlesoup/redis/SessionStore.kt
package io.github.kapu.turtlesoup.redis

import io.github.kapu.turtlesoup.config.JsonConfig
import io.github.kapu.turtlesoup.config.RedisConstants
import io.github.kapu.turtlesoup.config.RedisKeys
import io.github.kapu.turtlesoup.models.GameState
import io.github.kapu.turtlesoup.utils.RedisException
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.reactor.awaitSingle
import kotlinx.coroutines.reactor.awaitSingleOrNull
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import org.redisson.api.RedissonReactiveClient
import java.time.Duration
import org.redisson.client.RedisException as RedissonException

class SessionStore(
    private val redisson: RedissonReactiveClient,
) {
    private val json = JsonConfig.lenient

    suspend fun saveGameState(state: GameState) {
        val key = "${RedisKeys.SESSION}:${state.sessionId}"
        try {
            val value = json.encodeToString(state)
            val bucket = redisson.getBucket<String>(key)
            bucket.set(value, Duration.ofSeconds(RedisConstants.SESSION_TTL_SECONDS)).awaitSingleOrNull()

            logger.info { "game_state_saved session_id=${state.sessionId}" }
        } catch (e: RedissonException) {
            logger.error(e) { "game_state_save_failed session_id=${state.sessionId}" }
            throw RedisException("Failed to save game state: ${state.sessionId}", e)
        } catch (e: SerializationException) {
            logger.error(e) { "game_state_save_failed session_id=${state.sessionId}" }
            throw RedisException("Failed to save game state: ${state.sessionId}", e)
        }
    }

    suspend fun loadGameState(sessionId: String): GameState? {
        val key = "${RedisKeys.SESSION}:$sessionId"
        return try {
            val bucket = redisson.getBucket<String>(key)
            val value = bucket.get().awaitSingleOrNull() ?: return null
            json.decodeFromString<GameState>(value)
        } catch (e: RedissonException) {
            logger.error(e) { "game_state_load_failed session_id=$sessionId" }
            throw RedisException("Failed to load game state: $sessionId", e)
        } catch (e: SerializationException) {
            logger.error(e) { "game_state_load_failed session_id=$sessionId" }
            throw RedisException("Failed to load game state: $sessionId", e)
        }
    }

    suspend fun deleteSession(sessionId: String) {
        val sessionKey = "${RedisKeys.SESSION}:$sessionId"
        try {
            val bucket = redisson.getBucket<String>(sessionKey)
            bucket.delete().awaitSingleOrNull()
            logger.info { "session_deleted session_id=$sessionId" }
        } catch (e: RedissonException) {
            logger.error(e) { "session_delete_failed session_id=$sessionId" }
            throw RedisException("Failed to delete session: $sessionId", e)
        }
    }

    suspend fun sessionExists(sessionId: String): Boolean {
        val key = "${RedisKeys.SESSION}:$sessionId"
        return try {
            val bucket = redisson.getBucket<String>(key)
            bucket.isExists.awaitSingle()
        } catch (e: RedissonException) {
            logger.error(e) { "session_exists_check_failed session_id=$sessionId" }
            throw RedisException("Failed to check session existence: $sessionId", e)
        }
    }

    suspend fun refreshTtl(sessionId: String): Boolean {
        val sessionKey = "${RedisKeys.SESSION}:$sessionId"
        return try {
            val bucket = redisson.getBucket<String>(sessionKey)
            val result = bucket.expire(Duration.ofSeconds(RedisConstants.SESSION_TTL_SECONDS)).awaitSingle()
            if (result) {
                logger.debug { "ttl_refreshed session_id=$sessionId" }
            }
            result
        } catch (e: RedissonException) {
            logger.error(e) { "ttl_refresh_failed session_id=$sessionId" }
            throw RedisException("Failed to refresh TTL: $sessionId", e)
        }
    }

    companion object {
        private val logger = KotlinLogging.logger {}
    }
}

File: game-bot-go/internal/turtlesoup/redis/session_store.go
package redis

import (
	"context"
	"fmt"
	"log/slog"
	"time"

	"github.com/valkey-io/valkey-go"

	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/gamesession"
	tsconfig "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/config"
	tsmodel "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/model"
)

// SessionStore: 바다거북스프 게임의 진행 상태(퍼즐 정보, 참여자, 이력 등)를 Redis에 JSON 형태로 저장하고 관리하는 저장소
// 공통 gamesession.Store를 내부적으로 사용하여 CRUD 로직을 위임합니다.
type SessionStore struct {
	base *gamesession.Store[tsmodel.GameState]
}

// NewSessionStore: 새로운 SessionStore 인스턴스를 생성합니다.
func NewSessionStore(client valkey.Client, logger *slog.Logger) *SessionStore {
	return &SessionStore{
		base: gamesession.NewStore[tsmodel.GameState](client, logger, gamesession.Config{
			KeyFunc: sessionKey,
			TTL:     time.Duration(tsconfig.RedisSessionTTLSeconds) * time.Second,
		}),
	}
}

// SaveGameState: 현재 게임 상태 객체(GameState)를 JSON으로 직렬화하여 Redis에 저장합니다. (TTL 갱신 포함)
func (s *SessionStore) SaveGameState(ctx context.Context, state tsmodel.GameState) error {
	if err := s.base.Save(ctx, state.SessionID, state); err != nil {
		return fmt.Errorf("save game state: %w", err)
	}
	return nil
}

// LoadGameState: Redis에 저장된 JSON 데이터를 조회하여 GameState 객체로 역직렬화합니다.
// 데이터가 없거나 만료된 경우 nil을 반환합니다.
func (s *SessionStore) LoadGameState(ctx context.Context, sessionID string) (*tsmodel.GameState, error) {
	state, err := s.base.Load(ctx, sessionID)
	if err != nil {
		return nil, fmt.Errorf("load game state: %w", err)
	}
	return state, nil
}

// DeleteSession: 게임 종료 시 해당 게임 세션의 데이터를 Redis에서 영구 삭제합니다.
func (s *SessionStore) DeleteSession(ctx context.Context, sessionID string) error {
	if err := s.base.Delete(ctx, sessionID); err != nil {
		return fmt.Errorf("delete session: %w", err)
	}
	return nil
}

// SessionExists: 세션이 존재하는지 확인합니다.
func (s *SessionStore) SessionExists(ctx context.Context, sessionID string) (bool, error) {
	exists, err := s.base.Exists(ctx, sessionID)
	if err != nil {
		return false, fmt.Errorf("session exists: %w", err)
	}
	return exists, nil
}

// RefreshTTL: 세션의 TTL을 연장합니다.
func (s *SessionStore) RefreshTTL(ctx context.Context, sessionID string) (bool, error) {
	ok, err := s.base.RefreshTTL(ctx, sessionID)
	if err != nil {
		return false, fmt.Errorf("refresh ttl: %w", err)
	}
	return ok, nil
}
